atomix/cluster/src/test/java/io/atomix/raft/MemberJoinTest.java

```diff
@@ -0,0 +1,312 @@
+            new DefaultRaftMember(id2, Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(id3, Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(id4, Type.ACTIVE, Instant.now()));
+
+    assertThat(m1.cluster().getMembers()).containsExactlyInAnyOrderElementsOf(expected);
+    assertThat(m2.cluster().getMembers()).containsExactlyInAnyOrderElementsOf(expected);
+    assertThat(m3.cluster().getMembers()).containsExactlyInAnyOrderElementsOf(expected);
+    assertThat(m4.cluster().getMembers()).containsExactlyInAnyOrderElementsOf(expected);
+  }
+
+  @Test
+  void shouldCommitOnAllMembers(@TempDir final Path tmp) {
+    // given - a cluster with 3 members and one new member joining
+    final var id1 = MemberId.from("1");
+    final var id2 = MemberId.from("2");
+    final var id3 = MemberId.from("3");
+    final var id4 = MemberId.from("4");
+
+    final var m1 = createServer(tmp, createMembershipService(id1, id2, id3));
+    final var m2 = createServer(tmp, createMembershipService(id2, id1, id3));
+    final var m3 = createServer(tmp, createMembershipService(id3, id1, id2));
+    final var m4 = createServer(tmp, createMembershipService(id4, id1, id2, id3));
+
+    CompletableFuture.allOf(
+            m1.bootstrap(id1, id2, id3), m2.bootstrap(id1, id2, id3), m3.bootstrap(id1, id2, id3))
+        .join();
+    m4.join().join();
+
+    // when - appending a new entry
+    final var leader = awaitLeader(m1, m2, m3, m4);
+    final var index = appendEntry(leader).write().join();
+
+    // then - all members received the entry
+    Awaitility.await("All members have committed the entry")
+        .untilAsserted(
+            () ->
+                assertThat(List.of(m1, m2, m3, m4))
+                    .allSatisfy(
+                        server ->
+                            assertThat(server.getContext().getCommitIndex()).isEqualTo(index)));
+  }
+
+  @Test
+  void shouldRequireAdjustedQuorum(@TempDir final Path tmp) {
+    // given - a cluster with 3 members and two new members joining
+    final var id1 = MemberId.from("1");
+    final var id2 = MemberId.from("2");
+    final var id3 = MemberId.from("3");
+    final var id4 = MemberId.from("4");
+    final var id5 = MemberId.from("5");
+
+    final var m1 = createServer(tmp, createMembershipService(id1, id2, id3));
+    final var m2 = createServer(tmp, createMembershipService(id2, id1, id3));
+    final var m3 = createServer(tmp, createMembershipService(id3, id1, id2));
+    final var m4 = createServer(tmp, createMembershipService(id4, id1, id2, id3));
+    final var m5 = createServer(tmp, createMembershipService(id5, id1, id2, id3));
+
+    CompletableFuture.allOf(
+            m1.bootstrap(id1, id2, id3), m2.bootstrap(id1, id2, id3), m3.bootstrap(id1, id2, id3))
+        .join();
+
+    m4.join().join();
+    m5.join().join();
+
+    // when - no quorum possible because three out of five members are down
+    m1.shutdown().join();
+    m2.shutdown().join();
```

Comment:
:wrench: Here you can shutdown m3. if m1, m2 shutdown it is not possible to have quorum with the old members anyway.